sections/02_modular_functional_code.qmd

```diff
@@ -1,6 +1,49 @@
+
+When you have different and discrete workflows it is easy to decide how to turn them into functions (*e.g.,* one function to transform data and one to analyse) but there may be instances when you have three different analyses that share maybe 90% of the same code. What does one do then? It still makes sense to write one function that represent the shared code keep the 10% external to the function. 
+
+#### Be consistent
+
+As discussed in the [previous section](sections/01_structured_workflow.qmd) consistency is key to making your work easier for others to understand and follow. In terms of introducing functions into your workflow make sure that you are consistent with how you name and describe them. If you use term `init` in one function, you should use it everywhere; do not use `initialise` or `int` elsewhere. Because functions preform an action on an object an informative name will have some combination of `verb` + `object` select the order and remain consistent. Finally also remain consistent with which case you use; snake_case is usually best,
+
+::: {.callout-tip collapse="true"}
+## Example
+
+Lets say I have two functions one initialise an object as a number and the other a character. Bad practice would be to name them as follows:
+
+-   `int_char()`
+-   `numInit()`
+
+Instead opt for consistency in both how we choose to represent the 'initialise' action as well as the verb-object order:
+
+-   `init_num()`
+-   `init_char()`
+
+Note that there is no right and wrong in terms of the words, order, or case that we use, just make sure that it is consistent and that it is clearly documented.
+
+:::
+
+### Defensive programming
+
+Defensive programming is all about anticipating errors and writing robust code. The aim is to ensure that your code fails with well-defined errors, and rests on the idea that we expect our code to fail. By applying a defensive programming philosophy (and adding checks into the code), you can find unexpected behaviour sooner, and with more meaningful error messages. Although this means a bit more work at the start it will make debugging the code a lot easier at a later date. Although there are many ways to pre-empt a good place to start is 1) to write checks for cases that we know need to be met *e.g.,* ensuring that the types of the input are correct (is the input actually a number and not a character?) as well as testing the boundaries of our data (for example asserting that two dataframes are the same dimension) and 2) writing well defined error messages should those checks fail *e.g.,* if we test the dimensions of our input vs output dataframes and they do not match we have an error message telling us so. Note that (broadly) defensive programming is a way to test the behaviour of your code and not the *answer* per se - it will not pick up that you have specified the wrong underling process *e.g.,* adding instead of subtracting - that is a task for [testing](sections/03_testing_code.qmd).
```

Comment:
- Maybe we could put the points 1) and 2) in an ordered list instead of within the text? That might make it easier to read.
- I don't understand the last sentence: Do you mean the actual result/return value with "answer per se"? Right now I can't come up with a better way to put it, but I am not sure that "behaviour of your code vs. answer per se" is clear. I will think about, but maybe you also have an idea

Reply:
To your last point - I think it is something we can avoid all together if we combine the defensive programming with the testing section. Since testing is arguably a form of defensive programming but it is different to writing checks/asserting conditions. It is still a distinction to make but I think if we frame section 3 as defensive programming we can have checks and tests as sub sections and allow us to articulate this idea better. For now I am going to do some high level;/grammar changes and in a separate commit experiment with a restructure (which we can then easily revert)